astropy/io/fits/fitsrec.py

```diff
@@ -581,19 +581,15 @@ def __setitem__(self, key, value):
     def _ipython_key_completions_(self):
         return self.names
 
-    def copy(self, order="C"):
+    def copy(self):
```

Comment:
This changes API - perhaps not an issue?

Also wonder a bit more generally: if `fitsrec` keeps links to some original fits file, then wouldn't the whole file be copied? (This would have been true also with the `deepcopy(self.__dict__)`). It feels like a very large hammer...

Reply:
To be honest, I kept the deepcopy because of the comment, but it is a bit unclear to me why it was necessary.
The memory duplication seems caused by the `super().copy` plus the deepcopy.

I tried
```
    def copy(self, order="C"):
        new = super().copy(order=order)

        return new
``` 

And all tests passed, so if keeping the API unmodified is a priority and the deepcopy is not necessary, I can commit the version  above.

Reply:
How weird!  So, presumably even just removing `.copy()` would work... (... ... I just tried, all tests pass without the override of `copy()`).

This was all introduced in #1676, but unfortunately without explanation beyond the docstring. I must admit I don't understand why a deepcopy of weird attributes is any better than sharing them. But I think `__getitem__` makes clearer what in fact is the problem: it is all in `_converted`. My sense is that we can use the same technique as in `__getitem__` to get the copy; indeed,
```
# We copy the .view since that will ensure that a separate
# ColDefs is created, and _converted filled accordingly
# (see __getitem__).
new = super(self.__class__, self[...]).copy(order=order)
```
Or, maybe simpler,
```
        # Go through recarray to ensure we do not share items with self
        # that we do not want to share (see __getitem__).
        out = self.view(np.recarray).copy(order=order).view(type(self))
        out._uint = self._uint
        out._coldefs = ColDefs(self._coldefs)
        return out
```

Reply:
> How weird! So, presumably even just removing `.copy()` would work... (... ... I just tried, all tests pass without the override of `copy()`).

Right... I should have seen it...

I've chosen the simpler solution...

Reply:
Some additional info in https://github.com/spacetelescope/PyFITS/commit/9096a66efa99b5111aeaf7b3f9e2bb47a1c467f8:

> This commit contains a fix for a bug that hadn't revealed itself until
making these other changes (but was necessary to fix for all the tests
to pass).  This is specifically the fact that the `FITS_rec.copy()`
method (inherited from `ndarray.copy()`) did not perform a deep copy of
all of a FITS_rec's attributes, including the arrays in the `._convert`
list.  This meant that when copying a `FITS_rec` with the copy method
the new array would have its own copy of the *raw* data, but the
converted arrays would remain shared.  This is especially bad in the new
case where slices of *all* of the columns are stored in `._convert`.
This meant that any update to the new table updated the old table as
well (likewise they shared the same column definitions).  So now
`FITS_rec.copy()` is overridden to ensure that all of the table's
attributes get a deep copy.  This in turn revealed that since
`Column.__new__` expects a minimum number of arguments,
`copy.(deep)copy()` did not work on it without first defining custom
`__(deep)copy__` methods.
> 
> *whew!*

Handling of `_converted` columns is not simple (as far as I remember, but that's `io.fits` so), so maybe the comment helps a bit understanding the original issue. Are we sure that after changing copy's implementation the converted column are not shared with the original table ? (not sure if there is a test for that...)

Reply:
Thanks @saimn for the additional ref.
I'm going to look for a way to check about converted columns.